core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,873 @@
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+
+      questionPlayerEngineService.init(
+        multipleQuestionsBackendDict, successHandler, failHandler);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(questionPlayerEngineService.isAnswerBeingProcessed()).toBe(false);
+      expect(successCallback).toHaveBeenCalled();
+    });
+
+    it('should not submit answer again if the answer ' +
+      'is already being processed', () => {
+      let successCallback = jasmine.createSpy('success');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult');
+
+      questionPlayerEngineService.setAnswerIsBeingProcessed(true);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(successCallback).not.toHaveBeenCalled();
+    });
+
+    it('should show warning message if the feedback ' +
+      'html content is null', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', null, null, []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+      let alertsServiceSpy =
+        spyOn(alertsService, 'addWarning').and.returnValue();
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => html);
+
+      singleQuestionBackendDict.question_state_data
+        .interaction.default_outcome.feedback.html = null;
+      questionPlayerEngineService.init(
+        [singleQuestionBackendDict], successHandler, failHandler);
+
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(alertsServiceSpy)
+        .toHaveBeenCalledWith('Expression parsing error.');
```

Comment:
This is a weird message. What does it have to do with the feedback html being null? I don't see any expressions to parse, should we have a better message here?

(Note that, if the code is wrong, it is worth fixing it.)

Reply:
changed.